src/v/kafka/server/group_metadata.h

```diff
@@ -125,6 +256,48 @@ inline group_metadata_version read_metadata_version(request_reader& reader) {
     return group_metadata_version{reader.read_int16()};
 }
 
+class group_metadata_serializer {
+public:
+    struct key_value {
+        iobuf key;
+        std::optional<iobuf> value;
+    };
+    struct impl {
+        virtual group_metadata_type get_metadata_type(iobuf) = 0;
+        virtual key_value to_kv(group_metadata_kv) = 0;
+        virtual key_value to_kv(offset_metadata_kv) = 0;
+
+        virtual group_metadata_kv decode_group_metadata(model::record) = 0;
+        virtual offset_metadata_kv decode_offset_metadata(model::record) = 0;
+        virtual ~impl() = default;
+    };
```

Comment:
the abstraction layer is needed to support switching between implementations in runtime

Reply:
what I meant was: normal inheritance rather than pimpl would be much simpler. afaict we are never going to create another implementation of this. anyway, it's a minor point.

Reply:
actually we will create two implementations